### ISSUE TITLE
Testing issue

### ISSUE THREAD

**Issue by XiYY625** [hulianyuyy/Deep_Correlated_Prompting#6]
During training, the results on the training set can reach the performance reported in the paper. However, when evaluating the best-trained model on the test set, the performance drops significantly compared to both the training set and the results in the paper. How can this issue be resolved?

For example, at Epoch 4, global step 634, the validation metric (`val/the_metric`) reached **0.56246**, and the model was saved as the best checkpoint:

```
Epoch 4, global step 634: val/the_metric reached 0.56246 (best 0.56246), saving model to "result/exp_base_seed0/version_1/checkpoints/epoch=4-step=634.ckpt"
```

In the original paper, this corresponding metric was 56.21.

Then, I evaluated this model using the following command:

```bash
python run.py with data_root=./datasets/mmimdb num_gpus=1 num_nodes=1 per_gpu_batchsize=4  task_finetune_mmimdb load_path=result/exp_base_seed0/version_1/checkpoints/last.ckpt exp_name=mmimdb_experiment test_ratio=0.7 test_type=both test_only=True
```

The final results were:
```python
{
  'mmimdb/val/F1_Macro_epoch': tensor(0.4789, device='cuda:0'),
  'mmimdb/val/F1_Micro_epoch': tensor(0.5791, device='cuda:0'),
  'mmimdb/val/F1_Samples_epoch': tensor(0.5673, device='cuda:0'),
  'mmimdb/val/F1_Weighted_epoch': tensor(0.5783, device='cuda:0'),
  'mmimdb/val/loss': tensor(0.2296, device='cuda:0'),
  'mmimdb/val/loss_epoch': tensor(0.2548, device='cuda:0'),
  'val/the_metric': tensor(0.4789, device='cuda:0')
}
```
In comparison, the original paper reported 53.14.

**Comment by hulianyuyy** [comment]
Change the `last.chpt` in the command into `epoch=4-step=634.ckpt`.

**Comment by XiYY625** [comment]
Thank you for your reply.
When I run the command:

```bash
python run.py with data_root=./datasets/mmimdb num_gpus=1 num_nodes=1 per_gpu_batchsize=4 task_finetune_mmimdb load_path=result/mmimdb_experiment_seed0/version_0/checkpoints/last.ckpt exp_name=mmimdb_experiment test_ratio=0.7 test_type=both test_only=True simulate_missing=True
```

I get the following result:

```
DATALOADER:0 TEST RESULTS
{
 'mmimdb/val/F1_Macro_epoch': tensor(0.4391, device='cuda:0'),
 'mmimdb/val/F1_Micro_epoch': tensor(0.5313, device='cuda:0'),
 'mmimdb/val/F1_Samples_epoch': tensor(0.5158, device='cuda:0'),
 'mmimdb/val/F1_Weighted_epoch': tensor(0.5241, device='cuda:0'),
 'mmimdb/val/loss': tensor(0.4848, device='cuda:0'),
 'mmimdb/val/loss_epoch': tensor(0.4848, device='cuda:0'),
 'val/the_metric': tensor(0.4391, device='cuda:0')
}
```

Only the `F1_Micro_epoch` score (53.13) is close to the one reported in the paper (53.14).

**Comment by hulianyuyy** [comment]
You should use the checkpoint with the best performance for evaluation  i.e. epoch=4-step=634.ckpt, instead of using the checkpoint of the last epoch. Using the checkpoint of the last epoch would obviously leads to worse results.

**Comment by XiYY625** [comment]
> epoch=4-step=634.ckpt
Sorry for the confusion — the results from my first run were actually obtained using the checkpoint at epoch=4-step=634.ckpt.
The final results were:
{
  'mmimdb/val/F1_Macro_epoch': tensor(0.4789, device='cuda:0'),
  'mmimdb/val/F1_Micro_epoch': tensor(0.5791, device='cuda:0'),
  'mmimdb/val/F1_Samples_epoch': tensor(0.5673, device='cuda:0'),
  'mmimdb/val/F1_Weighted_epoch': tensor(0.5783, device='cuda:0'),
  'mmimdb/val/loss': tensor(0.2296, device='cuda:0'),
  'mmimdb/val/loss_epoch': tensor(0.2548, device='cuda:0'),
  'val/the_metric': tensor(0.4789, device='cuda:0')
}
In comparison, the original paper reported 53.14.
I encountered the same issue on other datasets as well.
Thank you for your reply.

**Comment by hulianyuyy** [comment]
I notice that 53.14 is the performance of missing image 70% in the paper. While in your experiments, you set `test_type=both` in the command. You should test with missing image to get the similar results.
